command/agent/http_test.go

```diff
@@ -61,6 +62,56 @@ func BenchmarkHTTPRequests(b *testing.B) {
 	})
 }
 
+// TestRootFallthrough tests rootFallthrough handler to
+// verify redirect and 404 behavior
+func TestRootFallthrough(t *testing.T) {
+	t.Parallel()
+
+	cases := []struct {
+		desc         string
+		path         string
+		expectedCode int
+	}{
+		{
+			desc:         "unknown endpoint 404s",
+			path:         "/v1/unknown/endpoint",
+			expectedCode: 404,
+		},
+		{
+			desc:         "root path redirects to ui",
+			path:         "/",
+			expectedCode: 307,
+		},
+	}
+
+	s := makeHTTPServer(t, nil)
+	defer s.Shutdown()
+
+	for _, tc := range cases {
+		t.Run(tc.desc, func(t *testing.T) {
+
+			reqURL := fmt.Sprintf("http://%s%s", s.Agent.config.AdvertiseAddrs.HTTP, tc.path)
+
+			// setup a client that doesn't follow redirects
+			client := &http.Client{
+				CheckRedirect: func(_ *http.Request, _ []*http.Request) error {
+					return http.ErrUseLastResponse
+				},
+			}
+
+			resp, err := client.Get(reqURL)
+			require.NoError(t, err)
+			require.Equal(t, tc.expectedCode, resp.StatusCode)
+
+			if tc.expectedCode == 307 {
+				loc, err := resp.Location()
+				require.NoError(t, err)
+				require.Equal(t, "/ui/", loc.Path)
```

Comment:
You can also have `expectedLocation` in test cases instead of special casing 307.

Reply:
In the case of a 404 there would be an error and no location header, I can at least change so it checks if the `tc.expectedPath != ""` so its driven by the tc and the assertion block isn't switching on a particular case